src/types/tuple.rs

```diff
@@ -201,54 +219,370 @@ impl PyTuple {
+    }
+
+    #[inline]
+    fn index<V>(&self, value: V) -> PyResult<usize>
+    where
+        V: ToPyObject,
+    {
+        self.as_sequence().index(value)
+    }
+
+    fn iter(&self) -> BoundTupleIterator<'py> {
+        BoundTupleIterator::new(self.clone())
+    }
+
+    fn iter_borrowed<'a>(&'a self) -> BorrowedTupleIterator<'a, 'py> {
+        BorrowedTupleIterator::new(Borrowed::from(self))
+    }
+
+    fn to_list(&self) -> Bound<'py, PyList> {
         self.as_sequence()
             .to_list()
             .expect("failed to convert tuple to list")
     }
 }
 
-index_impls!(PyTuple, "tuple", PyTuple::len, PyTuple::get_slice);
+impl<'a, 'py> Borrowed<'a, 'py, PyTuple> {
+    fn get_borrowed_item(self, index: usize) -> PyResult<Borrowed<'a, 'py, PyAny>> {
+        unsafe {
+            ffi::PyTuple_GetItem(self.as_ptr(), index as Py_ssize_t)
+                .assume_borrowed_or_err(self.py())
+        }
+    }
+
+    #[cfg(not(any(Py_LIMITED_API, PyPy)))]
+    unsafe fn get_borrowed_item_unchecked(self, index: usize) -> Borrowed<'a, 'py, PyAny> {
+        unsafe {
+            ffi::PyTuple_GET_ITEM(self.as_ptr(), index as Py_ssize_t).assume_borrowed(self.py())
+        }
+    }
+}
 
 /// Used by `PyTuple::iter()`.
-pub struct PyTupleIterator<'a> {
-    tuple: &'a PyTuple,
+pub struct PyTupleIterator<'a>(BorrowedTupleIterator<'a, 'a>);
+
+impl<'a> Iterator for PyTupleIterator<'a> {
+    type Item = &'a PyAny;
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        self.0.next().map(Borrowed::into_gil_ref)
+    }
+
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.0.size_hint()
+    }
+}
+
+impl<'a> DoubleEndedIterator for PyTupleIterator<'a> {
+    #[inline]
+    fn next_back(&mut self) -> Option<Self::Item> {
+        self.0.next_back().map(Borrowed::into_gil_ref)
+    }
+}
+
+impl<'a> ExactSizeIterator for PyTupleIterator<'a> {
+    fn len(&self) -> usize {
+        self.0.len()
+    }
+}
+
+impl FusedIterator for PyTupleIterator<'_> {}
+
+impl<'a> IntoIterator for &'a PyTuple {
+    type Item = &'a PyAny;
+    type IntoIter = PyTupleIterator<'a>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        PyTupleIterator(BorrowedTupleIterator::new(self.as_borrowed()))
+    }
+}
+
+/// Used by `PyTuple::into_iter()`.
+pub struct BoundTupleIterator<'py> {
+    tuple: Bound<'py, PyTuple>,
+    index: usize,
+    length: usize,
+}
+
+impl<'py> BoundTupleIterator<'py> {
+    fn new(tuple: Bound<'py, PyTuple>) -> Self {
+        let length = tuple.len();
+        BoundTupleIterator {
+            tuple,
+            index: 0,
+            length,
+        }
+    }
+}
+
+impl<'py> Iterator for BoundTupleIterator<'py> {
+    type Item = Bound<'py, PyAny>;
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index < self.length {
+            let item = unsafe {
+                BorrowedTupleIterator::get_item(Borrowed::from(&self.tuple), self.index).to_owned()
```

Comment:
I wonder if the amount of reuse could be increased by providing a method to go from `&'a mut BoundTupleIterator<'py>` to `&'a mut BorrowedTupleIterator<'a, 'py`>'?

I am torn on this though: While they have the same memory layout (after slapping on `repr(C)`), unsafe transmuting between feels a bit heavy-handed and the amount code duplication w.r.t. the indexing seems bearable.

Reply:
Yes I wondered about a similar thing; another option was to have `BoundTupleIterator` be 

```
struct BoundTupleIterator<'py> {
    tup: Bound<'py, PyTuple>,
    borrowed: BorrowedTupleIterator<'static, 'py, PyTuple>
```

where this type is sort of self-referencing; `tup` would own the reference used in `borrowed`, hence why `'static` lifetime on borrowed. It'd need some `unsafe` trickery again though, and in the end the argument of slightly more elegant code reuse didn't seem to justify unsafe here when I was able to formulate a safe solution that was only slightly clunky.

Reply:
Agreed. Let's life with this lack of reuse but the increase in safety it allows.